cmd/kubeadm/app/phases/controlplane/manifests.go

```diff
@@ -132,20 +132,22 @@ func CreateStaticPodFiles(manifestDir, patchesDir string, cfg *kubeadmapi.Cluste
 // getAPIServerCommand builds the right API server command from the given config object and version
 func getAPIServerCommand(cfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint *kubeadmapi.APIEndpoint) []string {
 	defaultArguments := map[string]string{
-		"advertise-address":               localAPIEndpoint.AdvertiseAddress,
-		"insecure-port":                   "0",
-		"enable-admission-plugins":        "NodeRestriction",
-		"service-cluster-ip-range":        cfg.Networking.ServiceSubnet,
-		"service-account-key-file":        filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPublicKeyName),
-		"client-ca-file":                  filepath.Join(cfg.CertificatesDir, kubeadmconstants.CACertName),
-		"tls-cert-file":                   filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerCertName),
-		"tls-private-key-file":            filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerKeyName),
-		"kubelet-client-certificate":      filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerKubeletClientCertName),
-		"kubelet-client-key":              filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerKubeletClientKeyName),
-		"enable-bootstrap-token-auth":     "true",
-		"secure-port":                     fmt.Sprintf("%d", localAPIEndpoint.BindPort),
-		"allow-privileged":                "true",
-		"kubelet-preferred-address-types": "InternalIP,ExternalIP,Hostname",
+		"advertise-address":                localAPIEndpoint.AdvertiseAddress,
+		"insecure-port":                    "0",
+		"enable-admission-plugins":         "NodeRestriction",
+		"service-cluster-ip-range":         cfg.Networking.ServiceSubnet,
+		"service-account-key-file":         filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPublicKeyName),
+		"service-account-issuer":           "https://kubernetes.default.svc",
```

Comment:
> I expected this to be the public address of the API server

if this should be the address, then we should set `service-account-issuer` to the same value in:
`"advertise-address":                localAPIEndpoint.AdvertiseAddress,`

would it be fine that separate instances of the kube-apiserver in an HA setup to have separate value for `service-account-issuer` (if this is based on the address)?

> I don't know how these default values propagate into something user-modifiable. cc @neolit123 for insight there

nowadays, kubeadm allows patching the manifest files before the kubelet starts on control-plane nodes (to create the static pods), so technically users can patch the manifests with the flag changes they want to. it's also possible to use "phases", i.e. create manifests, edit manifests with text editor, start the pods later.

Reply:
> would it be fine that separate instances of the kube-apiserver in an HA setup to have separate value for `service-account-issuer` (if this is based on the address)?

The issuer should be coherent between all servers in the cluster (or all servers should set an `--api-audiences` flag which includes all the servers' issuer names)

Reply:
hm, populating `--api-audiences` with addresses from subsequent api-servers is a problem. statically bootstrapping etcd server instances has a similar problem, yet the `--initial-cluster` of old members don't have to be updated with the addresses of new members that join later.

> The issuer should be coherent between all servers in the cluster

it would be best if we can use the same value of `--service-account-issuer` on all instances, for the default setup.

Reply:
issuer is used as `iss` claim in the token. it would be a URL. https://tools.ietf.org/html/rfc7519#section-4.1.1

> server names can change and people can make mistakes and want to fix them

users can specify `api-audiences` if they doing the audiences change.

also, will also allow users to pass in `service-account-issuer` and the default will still be `https://kubernetes.default.svc`.